server/lib/sync-agent.js

```diff
@@ -127,15 +128,48 @@ export default class SyncAgent {
       last_updated_at: options.last_updated_at || oneDayAgo.toISOString(),
       import_start_date: moment().subtract(this.ship.private_settings.import_days, "days").format()
     };
+
     const wrappedQuery = this.adapter.in.wrapQuery(query, replacements);
-    // Run the method for the specific adapter.
+      // Run the method for the specific adapter.
     return this.adapter.in.runQuery(this.client, wrappedQuery, options)
       .then(result => {
         this.adapter.in.closeConnection(this.client);
+
+        const validationResult = this.validateRows(result.rows);
+        if (!validationResult.isValid) {
+          return { entries: result.rows, errors: validationResult.errors };
+        }
+
+        if (this.db_type === "postgres" && this.postgresResultIsJson(result.fields)) {
+          return { entries: result.rows, errors: ["Column from postgres database is in json format"] };
+        }
+
         return { entries: result.rows };
       });
   }
 
+  postgresResultIsJson(columnNames) {
+    // 114 is special id for json type in postgres
+    return _.some(columnNames, (column) => column.dataTypeID === 114);
+  }
+
+  validateRows(result) {
+    if (this.db_type === "filesystem") return { isValid: true };
+    const errors = [];
+    const columnNames = Object.keys(result);
+
+    if (!_.includes(columnNames, "email") && !_.includes(columnNames, "external_id")) {
+      errors.push("Column names should include at least one required parameters: email or external_id");
+    }
+
+    if (_.some(columnNames, (field) => field.includes(".") || field.includes("$"))) {
+      errors.push("Column names should not contain special characters: `$`, `.`");
```

Comment:
@mkawalek are those errors actually displayed in the frontend ? I do not see any change to `src/index.js`